src/main/java/apoc/data/url/ExtractURL.java

```diff
@@ -0,0 +1,92 @@
+package apoc.data.url;
+
+import java.net.URL;
+import java.net.MalformedURLException;
+
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.UserFunction;
+
+/**
+ * This class is pretty simple.  It just constructs a java.net.URL instance
+ * from the user's input to do validation/parsing, and delegates the actual
+ * functionality to that class.  As such, the behavior of these functions
+ * matches that class, which is nice.
+ */
+public class ExtractURL {
+    /**
+     * In the context of a user function, we want to ignore parse errors.
+     */
+    protected static URL uriOrNull(String value) {
+        if (value == null) return null;
+        try {
+            URL u = new URL(value);
```

Comment:
create an URI instead?

Reply:
I tried URI first which is why the name here is leftover, but there are some odd tradeoffs with java.net.URI.   For example, in the case of the string "google.com", with a URI you get host="google.com" and path="google.com" which will surprise people, but is correct to the library since that URI is under-specified.   URIs also give you the ability to include URNs in addition to URLs, but I'm unsure how many people understand that since URNs are pretty uncommon these days.    This could be made more generic in order to use URIs rather than URLs though.